plugins/crypto/mbedtls/ua_pki_mbedtls.c

```diff
@@ -236,251 +238,158 @@ reloadCertificates(CertInfo *ci) {
-       ci->revocationListFolder.length == 0 &&
-       ci->certificateTrustList.raw.len == 0 &&
-       ci->certificateIssuerList.raw.len == 0 &&
-       ci->certificateRevocationList.raw.len == 0) {
-        UA_LOG_WARNING(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
-                       "PKI plugin unconfigured. Accepting the certificate.");
-        return UA_STATUSCODE_GOOD;
+static UA_Boolean
+mbedtlsCheckRevoked(CertInfo *ci, mbedtls_x509_crt *cert) {
+    for(mbedtls_x509_crl *crl = &ci->certificateRevocationList; crl; crl = crl->next) {
+        if(x509_name_eq(&cert->subject, &crl->issuer))
+            return (mbedtls_x509_crt_is_revoked(cert, crl) != 0);
     }
+    return false;
+}
 
-    /* Parse the certificate */
-    mbedtls_x509_crt remoteCertificate;
-
-    /* Temporary Object to parse the trustList */
-    mbedtls_x509_crt *tempCert = NULL;
-
-    /* Temporary Object to parse the revocationList */
-    mbedtls_x509_crl *tempCrl = NULL;
-
-    /* Temporary Object to identify the parent CA when there is no intermediate CA */
-    mbedtls_x509_crt *parentCert = NULL;
-
-    /* Temporary Object to identify the parent CA when there is intermediate CA */
-    mbedtls_x509_crt *parentCert_2 = NULL;
-
-    /* Flag value to identify if the issuer certificate is found */
-    int issuerKnown = 0;
-
-    /* Flag value to identify if the parent certificate found */
-    int parentFound = 0;
+#define UA_MBEDTLS_MAX_CHAIN_LENGTH 10
 
-    mbedtls_x509_crt_init(&remoteCertificate);
-    int mbedErr = mbedtls_x509_crt_parse(&remoteCertificate, certificate->data,
-                                         certificate->length);
-    if(mbedErr) {
-        /* char errBuff[300]; */
-        /* mbedtls_strerror(mbedErr, errBuff, 300); */
-        /* UA_LOG_WARNING(data->policyContext->securityPolicy->logger, UA_LOGCATEGORY_SECURITYPOLICY, */
-        /*                "Could not parse the remote certificate with error: %s", errBuff); */
-        return UA_STATUSCODE_BADSECURITYCHECKSFAILED;
+static UA_StatusCode
+mbedtls_verifyChain(CertInfo *ci, mbedtls_x509_crt *cert,
+                    mbedtls_x509_crt **old_issuers, int depth) {
+    /* Maxiumum chain length */
+    if(depth == UA_MBEDTLS_MAX_CHAIN_LENGTH)
+        return UA_STATUSCODE_BADCERTIFICATECHAININCOMPLETE;
+
+    /* Verification Step: Validity Period */
+    if(mbedtls_x509_time_is_future(&cert->valid_from) ||
+       mbedtls_x509_time_is_past(&cert->valid_to))
+        return UA_STATUSCODE_BADCERTIFICATETIMEINVALID;
+
+    /* Verification Step: Revocation Check */
+    if(mbedtlsCheckRevoked(ci, cert))
+        return UA_STATUSCODE_BADCERTIFICATEREVOKED;
+
+    /* Is the certificate in the trust list? If yes, then we are done. */
+    for(mbedtls_x509_crt *t = &ci->certificateTrustList; t; t = t->next) {
+        if(cert->tbs.len == t->tbs.len &&
+           memcmp(cert->tbs.p, t->tbs.p, cert->tbs.len) == 0)
+            return UA_STATUSCODE_GOOD;
     }
```

Comment:
It is understandable to shorten the verification, but two issues have to be considered here:

1. We cannot be sure that the TrustList is consistent. This is because of the RemoveCertificate() function which may eventually remove a CA certificate. Let's imagin the root CA has been removed (regardless of which sublist) and we currently look at an derived intermediate CA certificate which happens to be in the list of trusted certificates. Returning here now signals that the certificate chain is complete - but it isn't.
2. Expiration of CA certificates is not considered here. If we are not at the top of the chain here, it is possible that any CA the current certificate depends on might have been expired. Still we acknowledge the current certificate as valid - but it isn't.

To prevent point 1 one can consider to let RemoveCertificate() also remove all certificates that depend on the to-be-deleted one. Yet point 2 remains and seems to force us to completely validate the chain without a shortcut.

Reply:
1. That is correct. We might want to "walk up" until we reach a self-signed certificate.
It is not enough to have a trusted intermediate certificate.

2. This is considered. We check the expiration of every certificate in the chain already.

Reply:
2. If I understand the code correctly the expiration is only checked for each certificate that is processed. But if the shortcut applies (meaning that you do not walk up the chain any further) the expiration of all remaining certificates higher up the chain that is not checked anymore.

Reply:
We now walk the entire chain until the self-signed root CA.
Then we try if a trusted certificat is in the chain on the "way down".
Still we do the correct branching to explore the "tree of possible chains".